test/hotspot/jtreg/gc/TestReferenceRefersTo.java

```diff
@@ -197,8 +201,14 @@ private static void testConcurrentCollection() throws Exception {
             expectCleared(testPhantom1, "testPhantom1");
             expectCleared(testWeak2, "testWeak2");
             expectValue(testWeak3, testObject3, "testWeak3");
-            // This is true for all currently supported concurrent collectors.
-            expectNotCleared(testWeak4, "testWeak4");
+            // This is true for all currently supported concurrent collectors,
+            // except Shenandoah+IU, which allows clearing refs even when
+            // accessed during concurrent marking.
+            if (isShenandoahIUMode()) {
+              expectCleared(testWeak4, "testWeak4");
+            } else {
+              expectNotCleared(testWeak4, "testWeak4");
+            }
```

Comment:
I think I would prefer to keep this test program "generic", rather than having this Shenandoah IU mode intrusion.  So remove the old check of testWeak4 state here, and remove the check of obj4 below.  Instead, change the later check of testWeak4 being notified, where the new test is that either testWeak4 and obj4 are both null (IU and the like) or both are non-null (SATB and others).  Then add a couple of tests programs for the specific clearing or not clearing expected behaviors, with appropriate `@requires` restrictions.

Reply:
Right, that is even better. I made the base test generic, extracted the offending parts into its own test, and will push the Shenandoah specific test under a different PR. Thank you!